lib/context.js

```diff
@@ -0,0 +1,270 @@
+'use strict';
+
+const debug = require('debug')('koa-session:context');
+const Session = require('./session');
+const uid = require('uid-safe');
+const util = require('./util');
+
+const ONE_DAY = 24 * 60 * 60 * 1000;
+
+class ContextSession {
+  /**
+   * context session constructor
+   * @api public
+   */
+
+  constructor(ctx, opts) {
+    this.ctx = ctx;
+    this.opts = Object.assign({}, opts);
+    this.store = this.opts.store;
+  }
+
+  /**
+   * internal logic of `ctx.session`
+   * @return {Session} session object
+   *
+   * @api public
+   */
+
+  get() {
+    const session = this.session;
+    // already retrieved
+    if (session) return session;
+    // unset
+    if (session === false) return null;
+
+    // cookie session store
+    if (!this.store) this.initFromCookie();
+    return this.session;
```

Comment:
`session` is undefined here, `this.session` will be assigned after `this.initFromCookie()`